# Instacart Market Basket Analysis

### Student Name : Dikitso Motona
### Student Number : Bida24-328
### Course : Business Intelligence and Data Analytics
### Date: 03 / 03 / 2026


## Task 01 :  Project Setup and Data Understanding

## Step 1: Create Project Folder
Created `ADA2026_Instacart_Project` with folders `data/raw`, `data/interim`, `notebooks`, `outputs`, `docs`.

## Step 2: Downloading and Extracting the Dataset
The Instacart Market Basket Analysis dataset was downloaded from Kaggle and extracted into the `data/raw/instacart/` folder to preserve the original files.

## Step 3: Create the Notebook

A notebook named `T01_Project_Setup_and_Data_Understanding.ipynb` was created inside the `notebooks/` folder.  
Markdown headings will be used throughout the notebook to clearly separate and explain each step of the project.

### Step 4: Import Python Libraries
The following libraries are imported to support this project:
- **pandas** — for loading and manipulating data tables
- **numpy** — for numerical computations
- **matplotlib** — for data visualisation


In [2]:
import pandas as pd
import numpy as np
import matplotlib . pyplot as plt

print("Libraries imported successfully ✅")

Libraries imported successfully ✅


### Mount Google Drive
Mount Google Drive to access project files stored in the cloud.

In [3]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


### Step 5: Load the Data Tables
Load the main Instacart dataset tables into pandas dataframes from the `data/raw/instacart/` folder.

In [4]:
import pandas as pd

path = "/content/drive/MyDrive/ADA2026_Instacart_Project/Data/Raw/Instacart/"

orders = pd.read_csv( "/content/drive/MyDrive/ADA2026_Instacart_Project/Data/Raw/Instacart/"+ "orders.csv")
products = pd.read_csv( "/content/drive/MyDrive/ADA2026_Instacart_Project/Data/Raw/Instacart/"+ "products.csv")
order_products = pd.read_csv( "/content/drive/MyDrive/ADA2026_Instacart_Project/Data/Raw/Instacart/"+ "order_products__prior.csv")

The datasets `orders`, `products`, and `order_products` are now loaded into pandas dataframes and ready for analysis.

### Step 6: Inspect the Data
To verify that the datasets were loaded correctly, we display sample records, check column information and examine the shape of each table.

In [5]:

from IPython.display import display


print("=== Orders Dataset ===\n")
display(orders.head())
orders.info()
print("Shape:", orders.shape, "\n")


print("=== Products Dataset ===\n")
display(products.head())
products.info()
print("Shape:", products.shape, "\n")


print("=== Order_Products Dataset ===\n")
display(order_products.head())
order_products.info()
print("Shape:", order_products.shape)

=== Orders Dataset ===



,order_id,user_id,eval_set,order_number,order_dow,order_hour_of_day,days_since_prior_order
0,2539329,1,prior,1,2,8,NaN
1,2398795,1,prior,2,3,7,15.0
2,473747,1,prior,3,3,12,21.0
3,2254736,1,prior,4,4,7,29.0
4,431534,1,prior,5,4,15,28.0


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3421083 entries, 0 to 3421082
Data columns (total 7 columns):
 #   Column                  Dtype  
---  ------                  -----  
 0   order_id                int64  
 1   user_id                 int64  
 2   eval_set                object 
 3   order_number            int64  
 4   order_dow               int64  
 5   order_hour_of_day       int64  
 6   days_since_prior_order  float64
dtypes: float64(1), int64(5), object(1)
memory usage: 182.7+ MB
Shape: (3421083, 7) 

=== Products Dataset ===



,product_id,product_name,aisle_id,department_id
0,1,Chocolate Sandwich Cookies,61,19
1,2,All-Seasons Salt,104,13
2,3,Robust Golden Unsweetened Oolong Tea,94,7
3,4,Smart Ones Classic Favorites Mini Rigatoni Wit...,38,1
4,5,Green Chile Anytime Sauce,5,13


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 49688 entries, 0 to 49687
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   product_id     49688 non-null  int64 
 1   product_name   49688 non-null  object
 2   aisle_id       49688 non-null  int64 
 3   department_id  49688 non-null  int64 
dtypes: int64(3), object(1)
memory usage: 1.5+ MB
Shape: (49688, 4) 

=== Order_Products Dataset ===



,order_id,product_id,add_to_cart_order,reordered
0,2,33120,1,1
1,2,28985,2,1
2,2,9327,3,0
3,2,45918,4,1
4,2,30035,5,0


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32434489 entries, 0 to 32434488
Data columns (total 4 columns):
 #   Column             Dtype
---  ------             -----
 0   order_id           int64
 1   product_id         int64
 2   add_to_cart_order  int64
 3   reordered          int64
dtypes: int64(4)
memory usage: 989.8 MB
Shape: (32434489, 4)


### Data Inspection Summary Questions

- **Number of rows and columns:**
  - **Orders:** 3,421,083 rows × 7 columns  
  - **Products:** 49,688 rows × 4 columns  
  - **Order_Products:** 32,434,489 rows × 4 columns  

- **Columns that appear to be identifiers:**
  - `order_id` (present in Orders and Order_Products)  
  - `user_id` (Orders)  
  - `product_id` (Products and Order_Products)  

- **Missing values:**
`days_since_prior_order` in Orders contains missing values (`NaN`) —
    this is expected for a customer's very first order as there is no previous
    order to calculate the difference from.  
  - All other columns in Orders, Products and Order_Products appear to have complete data (no missing values)

 ### Step 7: Check for Missing Values
We check each dataset for missing values using `.isna().sum()`.  
- This counts the number of `NaN` entries per column.  
- Sorting in descending order highlights columns with the most missing values first.  

**Observations based on the outputs:**
- **Orders:** Only `days_since_prior_order` has missing values — **206,209 NaN entries**
  out of 3,421,083 rows (approximately 6% of the dataset). This is expected because
  a customer's very first order has no previous order to measure from.
- **Products:** No missing values detected across all 4 columns.
- **Order_Products:** No missing values detected across all 4 columns.

In [6]:
# Step 7: Check Missing Values
print("=== Missing Values in Orders Dataset ===\n")
orders_missing = orders.isna().sum().sort_values(ascending=False)
display(orders_missing)

print("\n=== Missing Values in Products Dataset ===\n")
products_missing = products.isna().sum().sort_values(ascending=False)
display(products_missing)

print("\n=== Missing Values in Order_Products Dataset ===\n")
order_products_missing = order_products.isna().sum().sort_values(ascending=False)
display(order_products_missing)

=== Missing Values in Orders Dataset ===



,0
days_since_prior_order,206209
user_id,0
order_id,0
eval_set,0
order_number,0
order_dow,0
order_hour_of_day,0



=== Missing Values in Products Dataset ===



,0
product_id,0
product_name,0
aisle_id,0
department_id,0



=== Missing Values in Order_Products Dataset ===



,0
order_id,0
product_id,0
add_to_cart_order,0
reordered,0


### Step 8: Data Dictionary
The data dictionary describes each column's meaning, data type, and key notes.  
It helps understand the dataset structure without inspecting raw files.

| Table          | Column                 | Meaning                                    | Data Type | Notes                     |
|----------------|------------------------|--------------------------------------------|-----------|---------------------------|
| Orders         | order_id               | Unique ID for each order                   | int64     | Primary key               |
| Orders         | user_id                | Unique ID for each user                    | int64     | Foreign key in Orders     |
| Orders         | order_number           | Sequence number of the order for the user  | int64     | User-specific ordering    |
| Orders         | order_dow              | Day of the week the order was placed       | int64     | 0=Sunday, 6=Saturday      |
| Orders         | days_since_prior_order | Days since prior order for same user       | float64   | NaN for first orders      |
| Orders         | order_hour_of_day      | Hour of the day the order was placed       | int64     | 0–23 (24hr format)        |
| Products       | product_id             | Unique ID for each product                 | int64     | Primary key               |
| Products       | product_name           | Name of the product                        | object    | Text description          |
| Products       | department_id          | Department category of the product        | int64     | Categorical ID            |
| Products       | aisle_id               | Aisle category of the product             | int64     | Links to aisles table     |
| Order_Products | add_to_cart_order      | Position of product in the cart           | int64     | Order of adding items     |
| Order_Products | reordered              | Whether product was reordered (1=yes,0=no) | int64     | Used for prediction       |

### Step 9: Save Interim Copies

We save interim copies of the datasets to the `Interim/` folder.  
This ensures we have backup CSVs of the loaded data for further analysis without modifying the original raw files.

In [ ]:

orders.to_csv("/content/drive/MyDrive/ADA2026_Instacart_Project/Data/Interim/orders_interim.csv", index=False)
products.to_csv("/content/drive/MyDrive/ADA2026_Instacart_Project/Data/Interim/products_interim.csv", index=False)
order_products.to_csv("/content/drive/MyDrive/ADA2026_Instacart_Project/Data/Interim/order_products_interim.csv", index=False)

print("✅ All interim files saved successfully in Data/Interim/")

## Step 10: Reflection

1. **What industry or business domain does this dataset relate to?**  
   The dataset is from a **grocery e-commerce business**, tracking customer orders, products and purchase patterns.

2. **How could this data help managers make decisions?**  
   - Optimize inventory and stock levels  
   - Plan marketing campaigns and promotions  
   - Recommend products to customers (cross-selling/upselling)  
   - Analyze customer behavior and loyalty  

3. **Which tables are useful for different types of data analysis?**  
   - **Descriptive Analysis:** `orders`, `products`, `order_products` — to summarize past sales and product trends  
   - **Predictive Analysis:** `orders`, `order_products` — to forecast future orders or reorder probability  
   - **Prescriptive Analysis:** `orders`, `order_products`, `products` — to suggest actions like personalized recommendations or inventory decisions